macOS test failure with LLVM >= 10

Even after you fix #478 , you will still get test failures on macOS with LLVM >= 10. Errors seem to look like:

```
/Applications/Xcode_12.2.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX11.0.sdk/usr/include/libkern/_OSByteOrder.h:97:1: error: unknown type name '__DARWIN_OS_INLINE'
```

I was hoping to revert to the macOS 10.15 SDK but it appears not to be installed on Github Actions.

Example failure: https://github.com/terralang/terra/runs/1751733849

Archive of the build log: 
[build_log.txt](https://github.com/terralang/terra/files/5858977/build_log.txt)

LLVM 10 got rid of the default definition of `__GNUC__`. I'm not quite sure why LLVM 10 isn't able to work with Apple's headers without this, but at any rate, the following reproducer demonstrates that putting this back solves the issue:

```
local C = terralib.includecstring [[
#ifndef __GNUC__
#define __GNUC__ 4
#endif
#include <stdlib.h>
]]
```

If you comment out the `#define` then you hit the error reported the original issue (reproduced below for clarity):

```
In file included from <buffer>:2:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/stdlib.h:66:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/wait.h:186:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/machine/endian.h:35:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/i386/endian.h:99:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/sys/_endian.h:130:
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/libkern/_OSByteOrder.h:97:1: error: unknown type name '__DARWIN_OS_INLINE'
__DARWIN_OS_INLINE
^
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/libkern/_OSByteOrder.h:98:9: error: expected ';' after top level declarator
uint16_t
        ^
        ;
```

I've been playing some more with the Clang frontend. I'm a bit confused because `-fgnuc-version` doesn't seem to be necessary in practice, but meanwhile I do see that `SDKROOT` does seem to matter now in Clang 11.

Contents of `test_gnuc.c`:

```c
#include <stdio.h>
int main(int argc, char *argv[]) {
#ifdef __GNUC__
  int gnuc = __GNUC__;
  printf("yep %d\n", gnuc);
#else 
  printf("nope.\n");
#endif
  return 0;
}
```

Apple Clang:

```$ clang test_gnuc.c -o test_gnuc
$ ./test_gnuc 
yep 4
$ which clang
/usr/bin/clang
$ clang --version
Apple clang version 12.0.0 (clang-1200.0.32.29)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

LLVM Clang 9 from binary:

```
$ ./llvm/llvm-9/clang+llvm-9.0.1-x86_64-apple-darwin/bin/clang test_gnuc.c -o test_gnuc
test_gnuc.c:1:10: fatal error: 'stdio.h' file not found
#include <stdio.h>
         ^~~~~~~~~
1 error generated.
$ ./llvm/llvm-9/clang+llvm-9.0.1-x86_64-apple-darwin/bin/clang test_gnuc.c -o test_gnuc -I "$(xcrun --sdk macosx --show-sdk-path)/usr/include"
...
(lots of warnings, but compiles)
$ ./test_gnuc 
yep 4
```

LLVM Clang 11 from binary:

```
$ ./llvm/llvm-11/clang+llvm-11.0.0-x86_64-apple-darwin/bin/clang test_gnuc.c -o test_gnuc
test_gnuc.c:1:10: fatal error: 'stdio.h' file not found
#include <stdio.h>
         ^~~~~~~~~
1 error generated.
$ ./llvm/llvm-11/clang+llvm-11.0.0-x86_64-apple-darwin/bin/clang test_gnuc.c -o test_gnuc -I "$(xcrun --sdk macosx --show-sdk-path)/usr/include"
...
(lots of warnings, then one error):
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/stdio.h:220:5: error: 'TARGET_OS_IPHONE' is not defined, evaluates to 0 [-Werror,-Wundef-prefix=TARGET_OS_]
#if TARGET_OS_IPHONE
    ^
$ SDKROOT="$(xcrun --sdk macosx --show-sdk-path)" ./llvm/llvm-11/clang+llvm-11.0.0-x86_64-apple-darwin/bin/clang test_gnuc.c -o test_gnuc -I "$(xcrun --sdk macosx --show-sdk-path)/usr/include"
$ ./test_gnuc 
yep 4
```

This last bit is interesting... I wasn't expecting that part. Maybe I need to fiddle with the `SDKROOT` value for Terra. That didn't used to matter, so I'm a bit surprised.

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -44,7 +44,7 @@ jobs:
             static: '0'
           - os: 'macos-10.15'
             slib: '0'
-          # LLVM 11 binaries are broken on recent macOS
+          # LLVM 10 binaries are broken on recent macOS
           - os: 'macos-10.15'
             llvm: '10'
 

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -44,6 +44,9 @@ jobs:
             static: '0'
           - os: 'macos-10.15'
             slib: '0'
+          # LLVM 11 binaries are broken on recent macOS
+          - os: 'macos-10.15'
+            llvm: '10'
 
           # CMake
           - llvm: '3.5'

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -44,11 +44,6 @@ jobs:
             static: '0'
           - os: 'macos-10.15'
             slib: '0'
-          # FIXME: https://github.com/terralang/terra/issues/479
-          - os: 'macos-10.15'
-            llvm: '10'
-          - os: 'macos-10.15'
-            llvm: '11'
 
           # CMake
           - llvm: '3.5'

--- a/release/share/terra/README.md
+++ b/release/share/terra/README.md
@@ -145,8 +145,8 @@ The current recommended version of LLVM is **6.0**. The following versions are a
 | 7.0 | :heavy_check_mark: | :heavy_check_mark: | requires CMake |
 | 8.0 | :heavy_check_mark: | :heavy_check_mark: | |
 | 9.0 | :heavy_check_mark: | :heavy_check_mark: | |
-| 10.0 | :heavy_check_mark: | :heavy_check_mark: | [#476](https://github.com/terralang/terra/issues/476) [#477](https://github.com/terralang/terra/issues/477) [#479](https://github.com/terralang/terra/issues/479) |
-| 11.0 | :heavy_check_mark: | :heavy_check_mark: | [#476](https://github.com/terralang/terra/issues/476) [#477](https://github.com/terralang/terra/issues/477) [#479](https://github.com/terralang/terra/issues/479) |
+| 10.0 | :heavy_check_mark: | :heavy_check_mark: | [#476](https://github.com/terralang/terra/issues/476) [#477](https://github.com/terralang/terra/issues/477) |
+| 11.0 | :heavy_check_mark: | :heavy_check_mark: | [#476](https://github.com/terralang/terra/issues/476) [#477](https://github.com/terralang/terra/issues/477) |
 
 ### Instructions for Building LLVM from Source
 

--- a/src/terralib.lua
+++ b/src/terralib.lua
@@ -3426,7 +3426,10 @@ function terra.includecstring(code,cargs,target)
     	args:insert("-internal-isystem")
     	args:insert(path)
     end
-    
+    if ffi.os == "Darwin" and terralib.llvm_version >= 100 then
+      args:insert("-fgnuc-version=4")
+    end
+
     if cargs then
         args:insertall(cargs)
     end

--- a/src/terralib.lua
+++ b/src/terralib.lua
@@ -3426,7 +3426,14 @@ function terra.includecstring(code,cargs,target)
     	args:insert("-internal-isystem")
     	args:insert(path)
     end
-    if ffi.os == "Darwin" and terralib.llvm_version >= 100 then
+    -- Obey the SDKROOT variable on macOS to match Clang behavior.
+    local sdkroot = os.getenv("SDKROOT")
+    if sdkroot then
+      args:insert("-isysroot")
+      args:insert(sdkroot)
+    end
+    -- Set GNU C version to match values observed on macOS.
+    if ffi.os == "OSX" and terralib.llvm_version >= 100 then
       args:insert("-fgnuc-version=4")
     end
 

--- a/travis.sh
+++ b/travis.sh
@@ -174,7 +174,7 @@ if [[ $(uname) = Darwin ]]; then
 
   # workaround for https://github.com/terralang/terra/issues/365
   if [[ ! -e /usr/include ]]; then
-    export INCLUDE_PATH="$(xcrun --sdk macosx --show-sdk-path)/usr/include"
+    export SDKROOT="$(xcrun --sdk macosx --show-sdk-path)"
   fi
 
   export PATH=$PWD:$PATH

